src/core/batch.rkt

```diff
@@ -82,71 +82,173 @@
+       (define mult-idx (append-node `(* ,(hash-ref mappings args) ,(hash-ref mappings args))))
+       (define one-idx (append-node 1))
+       (define add-idx (append-node `(+ ,mult-idx ,one-idx)))
+       (define sqrt-idx (append-node `(sqrt ,add-idx)))
+       (define add2-idx (append-node `(+ ,(hash-ref mappings args) ,sqrt-idx)))
+       (hash-set! mappings n (append-node `(log ,add2-idx)))]
+      [(list 'acosh args)
+       (define mult-idx (append-node `(* ,(hash-ref mappings args) ,(hash-ref mappings args))))
+       (define -one-idx (append-node -1))
+       (define add-idx (append-node `(+ ,mult-idx ,-one-idx)))
+       (define sqrt-idx (append-node `(sqrt ,add-idx)))
+       (define add2-idx (append-node `(+ ,(hash-ref mappings args) ,sqrt-idx)))
+       (hash-set! mappings n (append-node `(log ,add2-idx)))]
+      [(list 'atanh args)
+       (define neg-idx (append-node `(neg ,(hash-ref mappings args))))
+       (define one-idx (append-node 1))
+       (define add-idx (append-node `(+ ,one-idx ,(hash-ref mappings args))))
+       (define sub-idx (append-node `(+ ,one-idx ,neg-idx)))
+       (define div-idx (append-node `(/ ,add-idx ,sub-idx)))
+       (define log-idx (append-node `(log ,div-idx)))
+       (define half-idx (append-node 1/2))
+       (hash-set! mappings n (append-node `(* ,half-idx ,log-idx)))]
+      [(list op args ...)
+       (hash-set! mappings n (append-node (cons op (map (curry hash-ref mappings) args))))]
+      [(approx spec impl) (hash-set! mappings n (append-node (approx spec (hash-ref mappings impl))))]
+      [_ (hash-set! mappings n (append-node node))]))
+
+  (define roots* (vector-map (curry hash-ref mappings) roots))
+  (define nodes* (list->vector (reverse icache)))
+
+  ; This may be too expensive to handle simple 1/2, 1/3 and 2/3 zombie nodes..
+  #;(remove-zombie-nodes (batch nodes* roots* vars (vector-length nodes*)))
+
+  (batch nodes* roots* vars (vector-length nodes*)))
+
+; The function removes any zombie nodes from batch
+(define (remove-zombie-nodes input-batch)
+  (define nodes (batch-nodes input-batch))
+  (define roots (batch-roots input-batch))
+  (define nodes-length (batch-nodes-length input-batch))
+
+  (define zombie-mask (make-vector nodes-length #t))
+  (for ([root (in-vector roots)])
+    (vector-set! zombie-mask root #f))
+  (for ([node (in-vector nodes (- nodes-length 1) -1 -1)]
+        [zmb (in-vector zombie-mask (- nodes-length 1) -1 -1)]
+        #:when (not zmb))
+    (match node
+      [(list op args ...) (map (λ (n) (vector-set! zombie-mask n #f)) args)]
+      [(approx spec impl) (vector-set! zombie-mask impl #f)]
+      [_ void]))
+
+  (define mappings (make-hash (map (λ (n) (cons n n)) (build-list nodes-length values))))
+
+  (define nodes* '())
+  (for ([node (in-vector nodes)]
+        [zmb (in-vector zombie-mask)]
+        [n (in-naturals)])
+    (if zmb
+        (map (λ (n) (hash-set! mappings n (sub1 (hash-ref mappings n)))) (range n nodes-length))
```

Comment:
I don't really care but technically if you write this as a `for` loop it'll be faster because you won't materialize the `range` as a list.